examples/qwen3_vl/main.zig

```diff
@@ -0,0 +1,745 @@
+    const factor = preprocessor_config.patch_size * preprocessor_config.temporal_patch_size;
+    const min_pixels = preprocessor_config.size.shortest_edge;
+    const max_pixels = preprocessor_config.size.longest_edge;
+    stdx.debug.assert(@max(height, width) / @min(height, width) <= 200, "Invalid image ratio", .{});
+
+    // Calculate the resized height and width
+    var h_resized = @round(stdx.math.divFloat(f64, height, factor)) * @as(f64, @floatFromInt(factor));
+    var w_resized = @round(stdx.math.divFloat(f64, width, factor)) * @as(f64, @floatFromInt(factor));
+    if (h_resized * w_resized > max_pixels) {
+        const beta = std.math.sqrt(@as(f64, @floatFromInt(height * width)) / max_pixels);
+        h_resized = @max(@as(f64, @floatFromInt(factor)), std.math.floor(@as(f64, @floatFromInt(height)) / beta / @as(f64, @floatFromInt(factor))) * @as(f64, @floatFromInt(factor)));
+        w_resized = @max(@as(f64, @floatFromInt(factor)), std.math.floor(@as(f64, @floatFromInt(width)) / beta / @as(f64, @floatFromInt(factor))) * @as(f64, @floatFromInt(factor)));
+    } else if (h_resized * w_resized < min_pixels) {
+        const beta = std.math.sqrt(min_pixels) / @as(f64, @floatFromInt(height * width));
+        h_resized = @max(@as(f64, @floatFromInt(factor)), std.math.ceil(@as(f64, @floatFromInt(height)) * beta / @as(f64, @floatFromInt(factor))) * @as(f64, @floatFromInt(factor)));
+        w_resized = @max(@as(f64, @floatFromInt(factor)), std.math.ceil(@as(f64, @floatFromInt(width)) * beta / @as(f64, @floatFromInt(factor))) * @as(f64, @floatFromInt(factor)));
+    }
+    const patch_size = config.vision_config.patch_size;
+
+    // Calculate the number of image pad tokens
+    const number_image_pad_tokens = 1 * (@as(u32, @intFromFloat(h_resized)) / patch_size) * (@as(u32, @intFromFloat(w_resized)) / patch_size) / std.math.pow(u32, config.vision_config.spatial_merge_size, 2);
+
+    // Apply the chat template to the prompt
+    const result = try applyChatTemplate(allocator, tokenizer, prompt, number_image_pad_tokens);
+    const prompt_encoded = result.prompt_tokens;
+    const prompt_shape = result.prompt_shape;
+
+    defer allocator.free(prompt_encoded);
+    const prompt_buffer = try allocator.alloc(u32, max_seq_len);
+
+    // Create the HostBuffers for the prompt, prompt shape, image size and token index
+    @memcpy(prompt_buffer[0..prompt_encoded.len], prompt_encoded);
+    const prompt_tokens = zml.HostBuffer.fromSlice(zml.Shape.init(.{ .bs = 1, .seq = max_seq_len }, .u32), prompt_buffer);
+    const prompt_shape_buffer = try zml.HostBuffer.empty(allocator, zml.Shape.init(.{ .chw = 3 }, .i32));
+    @memcpy(prompt_shape_buffer.mutItems(i32), &prompt_shape);
+    const image_size_buffer = try zml.HostBuffer.empty(allocator, zml.Shape.init(.{ .chw = 3 }, .i32));
+    @memcpy(image_size_buffer.mutItems(i32), &image_size);
+    const token_index_buffer = try zml.HostBuffer.empty(allocator, zml.Shape.init(.{}, .i64));
```

Comment:
feels like we should have a `fromArray` that allocates and copy the array content and rename current `fromArray` to `fromArrayPtr`